### PR TITLE
Prototype mt_setup

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
All of ModelingToolkit was made just as standard Julia functions and types. This is nice because then it's very interactive, but sometimes you want to do something at compile time. This idea is a simple macro `@mt_setup` which replaces things (thing) that eval with the piece that generates the code. In the macro form, this then is transformed at compile time to just be a code generating macro from MT syntax.

To really make this useful, we'll want to make it so that way more than 1 function is allowed. For example, generate the ODE function and its Jacobian, and then build the ODEFunction (or ODEProblem). Maybe we could have a bit more DSL to it, but I kind of like that you can copy-paste what worked in the global scope and now stick it in a function with this. That's not true without the macro because of the evals.

### CODE REVIEW

File: src/mt_setup.jl [hunk 1]
```diff
@@ -0,0 +1,6 @@
+macro mt_setup(expr)
+    MacroTools.postwalk(expr) do x
+        x == :ODEFunction ? :generate_function : x
```
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/mt_setup.jl]: yes it is. So then do we document `generate_function` so that way they can use the macro?

**Comment by ChrisRackauckas** [comment]
The main purpose of this is because MT cannot be used inside of a function right now because of the eval. We need some way to just throw a quick macro on it to make it not tied to the global scope.

File: src/mt_setup.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (HarrisonGrodin) [reply src/mt_setup.jl]: `generate_function` is already exported. Which macro - `@mt_setup`?

**Comment by HarrisonGrodin** [comment]
Couldn't this be solved by making `generate_function` a higher-order function which returns a function, rather than an expression? We could even just use an `@eval` before returning from `generate_function`, I believe.

**Comment by ChrisRackauckas** [comment]
The purpose is to not `eval` at all.

**Comment by HarrisonGrodin** [comment]
It seems like `@mt_setup` is just the no-op macro, since replacing `ODEFunction` with `generate_function` isn't necessary:
```julia
macro mt_setup(expr)
    expr
end
```
Rather than supporting an additional API for something more general than ModelingToolkit, it seems like this should exist elsewhere (`Base`, a utility package, etc.).

File: src/mt_setup.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/mt_setup.jl]: what about doing this like a trait system, where functions can opt in to have a change in here? I think we want to keep the copy-pastability into the macro.

**Comment by ChrisRackauckas** [comment]
I think we are going to want something like this, but not with the hardcoded name change. I'll get an update to this which is more extensible for the next release, and then we can keep tweaking it.

**Pull request closed without merging**